src/main/java/org/gaul/s3proxy/EncryptedBlobStore.java

```diff
@@ -0,0 +1,784 @@
+                String[] ranges = range.split("-", 2);
+                if (ranges[0].isEmpty()) {
+                    // clear current ranges to avoid multiple ranges
+                    getOptions.getRanges().clear();
+
+                    // handle to read from the end a specific amount of bytes
+                    long length = Long.parseLong(ranges[1]);
+                    decryption =
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, 0L, length);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.tail(length);
+                    } else {
+                        long startAt = decryption.calculateTail();
+                        getOptions.startAt(startAt);
+                    }
+                } else if (ranges[1].isEmpty()) {
+                    // clear current ranges to avoid multiple ranges
+                    getOptions.getRanges().clear();
+
+                    // handle to read from an offset till the end
+                    long offset = Long.parseLong(ranges[0]);
+                    decryption =
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, offset, -1L);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.startAt(offset);
+                    } else {
+                        long startAt = decryption.getStartAt();
+                        getOptions.startAt(startAt);
+                    }
+                } else {
+                    // clear current ranges to avoid multiple ranges
+                    getOptions.getRanges().clear();
+
+                    // handle to read from an offset a specific amount of bytes
+                    long offset = Long.parseLong(ranges[0]);
+                    long end = Long.parseLong(ranges[1]);
+                    long length = end - offset + 1;
+                    decryption =
+                        new Decryption(secretKey, delegate(), containerName,
+                            blobName, offset, length);
+
+                    // if the blob is not encrypted we will fall back
+                    if (!decryption.isEncrypted()) {
+                        getOptions.range(offset, end);
+                    } else {
+                        long startAt = decryption.getStartAt();
+                        long endAt = decryption.calculateEndAt(end);
+                        getOptions.range(startAt, endAt);
+                    }
+                }
+            }
+
+            // initialize the decryption if no ranges were requested
+            if (decryption == null) {
+                decryption =
+                    new Decryption(secretKey, delegate(), containerName,
```

Comment:
Can we sink the instantiation of `Decryption` into `decryptBlob` and update its signature to:
`private Blob decryptBlob(String container, Blob blob)`? I think the `range` calculation is the only difference and maybe it could actually be done in the `Decryption` constructor itself (via passing the `Blob`, rather than just the name)?

Reply:
I was also thinking about this but we need to adjust ```getOptions``` before it get passed to the delegate to get the blob.